### DISCUSSION TITLE
How many groups do you have in your realm and how they look like in terms of levels of hierarchy?

### DISCUSSION THREAD

**Discussion by pedroigor** [keycloak/keycloak#31122] [Feedback]
Hi,

Can you share how you are using groups today based on the following points?

* How many groups do you have in the database?
* How many top-level groups do you have in a realm?
* How many levels do you have in the hierarchy, approximately?
* How many sub-groups do you have in each level of the hierarchy, approximately?

Helping with this query should help us to improve the scalability of groups and organizations in Keycloak.

Thanks in advance.

**Comment by jbman** [comment]
In the largest setup, 43.000 top-level groups are in the database. Sub-groups aren't used in this setup.
Each group represents a company and some employees of the company are members of the group (usually only a small number of users in the group, like 1-50 users). 
The users of one company do not share same email domain and do not use a company specific IdP for login (all users use same IdP).

**Comment by thomasdarimont** [comment]
I have already seen various ways of using groups.

Single-tenant realm with coarse groups to simplify role assignment (admins/support/guests/partners) with ~10 groups for simple role assignment to users.

Multi-tenant realm with coarse-grained groups to express tenant membership with ~10s up to 10k tenant groups. Sometimes we see additional subgroups to differentiate users within a tenant group, e.g.

- /tenants/tenant1/Owner
- /tenants/tenant1/Admin
- /tenants/tenant1/user
- /tenants/tenant1/guest

This structure makes it possible to invite users from other tenants with different roles, e.g. a user from /tenants/tenant1/user can be a member of /tenants/tenant2/guest.

Furthermore, these tenant groups have descriptive attributes like a TenantId and tenant specific policy configurations, e.g. the booked service level (basic, silver, gold) and the date of tenant creation, as well as information when which terms were accepted.
I think this model can infact be replaced with the new support for organizations.

Sometimes I also see that groups are used to differentiate between different user classes:

- /staff
- /admin
- /partner
- /students
- /auditors
- /external

There are also scenarios in which groups are used to map authorisations in different areas of a product suite, which can extend across several clients.

- /products/product1/context1/admin
- /products/product1/context2/ordermanager
- /products/product2/context3/director

Also when Keycloak Groups are used in combination with LDAP or other sources for groups (like IdP brokering etc.) I see multiple group hiearchies being used. One group hierarchy is then managed via LDAP / IdP brokering, the other managed via Keycloaks Admin UI / API.

**Comment by vaish0117** [comment]
@pedroigor 
In v26.0.0, how many top level groups can we create in single realm without affecting performance?
We need to create one/two top level groups per customer organization, and each customer group can have many users. (assuming maximum of 2-3 sub-groups).
We have CRUD operations to get user-group associations etc.
Is performance and scale something to be worried about? 

Another question, 
How many top level groups can I create in a single realm as of today? (assuming no sub-groups)
